See also this warning please:
```
[1161/1660] Compiling C object librz/core/librz_core.so.0.3.0-git.p/task.c.o
In file included from ../librz/core/task.c:5:
In file included from ../librz/include/rz_core.h:34:
../librz/include/rz_heap_glibc.h:34:9: warning: 'MAX' macro redefined [-Wmacro-redefined]
#define MAX(a, b)              (((a) > (b)) ? (a) : (b))
        ^
/usr/include/sys/param.h:315:9: note: previous definition is here
#define MAX(a,b) (((a)>(b))?(a):(b))
        ^
1 warning generated.
```
Just use `RZ_MAX`

--- a/librz/core/cmd_linux_heap_glibc.c
+++ b/librz/core/cmd_linux_heap_glibc.c
@@ -51,3 +51,13 @@ RZ_IPI int rz_cmd_heap_fastbins_print(void *data, const char *input) {
 RZ_IPI RzCmdStatus rz_cmd_heap_arena_bins_print_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
 	call_handler(rz_cmd_heap_arena_bins_print_handler, argc, argv, mode);
 }
+
+/* API functions for Heap Viewer in Cutter */
+
+RZ_API RzList *rz_heap_arenas_list(RzCore *core) {
+	call_handler(rz_heap_arena_list_wrapper);
+}
+
+RZ_API RzList *rz_heap_chunks_list(RzCore *core, ut64 m_arena) {
+	call_handler(rz_heap_chunks_list_wrapper, m_arena);
+}

--- a/librz/core/linux_heap_glibc.c
+++ b/librz/core/linux_heap_glibc.c
@@ -1252,14 +1252,7 @@ void GH(print_malloc_states)(RzCore *core, GHT m_arena, MallocState *main_arena,
 		while (GH(is_arena)(core, m_arena, ta->GH(next)) && ta->GH(next) != m_arena) {
 			ut64 ta_addr = ta->GH(next);
 			if (!GH(rz_heap_update_main_arena)(core, ta->GH(next), ta)) {
-				if (json) {
-					pj_end(pj);
-					pj_end(pj);
-					rz_cons_println(pj_string(pj));
-					pj_free(pj);
-				}
-				free(ta);
-				return;
+				goto end;
 			}
 			if (!json) {
 				rz_cons_printf("Thread arena(addr=");
@@ -1291,6 +1284,7 @@ void GH(print_malloc_states)(RzCore *core, GHT m_arena, MallocState *main_arena,
 			}
 		}
 	}
+end:
 	if (json) {
 		pj_end(pj);
 		pj_end(pj);
@@ -1964,16 +1958,10 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 	char *top_title = NULL, *top_data = NULL, *node_title = NULL, *node_data = NULL;
 	bool first_node = true;
 	top_data = rz_str_new("");
+	RzList *chunks = GH(rz_heap_chunks_list)(core, main_arena, m_arena, m_state);
 	if (mode == RZ_OUTPUT_MODE_JSON) {
 		if (!pj) {
-			rz_cons_newline();
-			free(g);
-			free(top_data);
-			free(main_arena);
-			rz_cons_canvas_free(can);
-			rz_config_hold_restore(hc);
-			rz_config_hold_free(hc);
-			return RZ_CMD_STATUS_ERROR;
+			goto end;
 		}
 		pj_o(pj);
 		pj_ka(pj, "chunks");
@@ -1989,7 +1977,6 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 		rz_agraph_set_title(g, "Heap Layout");
 		top_title = rz_str_newf("Top chunk @ 0x%" PFMT64x "\n", (ut64)main_arena->GH(top));
 	}
-	RzList *chunks = GH(rz_heap_chunks_list)(core, main_arena, m_arena, m_state);
 	rz_list_foreach (chunks, iter, pos) {
 		if (mode == RZ_OUTPUT_MODE_STANDARD || mode == RZ_OUTPUT_MODE_LONG) {
 			GH(print_heap_chunk_simple)
@@ -2021,13 +2008,9 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 			rz_cons_printf("f %s %d 0x%" PFMT64x "\n", name, (int)pos->size, (ut64)pos->addr);
 			free(name);
 		} else if (mode == RZ_OUTPUT_MODE_LONG_JSON) { // graph
-			if (node_title) {
-				free(node_title);
-			}
+			free(node_title);
+			free(node_data);
 			node_title = rz_str_newf("  Malloc chunk @ 0x%" PFMT64x " ", (ut64)pos->addr);
-			if (node_data) {
-				free(node_data);
-			}
 			node_data = rz_str_newf("size: 0x%" PFMT64x " status: %s\n", (ut64)pos->size, pos->status);
 			chunk_node = rz_agraph_add_node(g, node_title, node_data);
 			if (first_node) {
@@ -2067,6 +2050,7 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 		}
 		rz_agraph_print(g);
 	}
+end:
 	rz_cons_newline();
 	free(g);
 	free(top_data);

--- a/librz/core/linux_heap_glibc.c
+++ b/librz/core/linux_heap_glibc.c
@@ -1388,7 +1388,7 @@ RZ_API RzList *GH(rz_heap_bin_content_list)(RzCore *core, MallocState *main_aren
 	GH(get_brks)
 	(core, &brk_start, &brk_end);
 	if (brk_start == GHT_MAX || brk_end == GHT_MAX) {
-		eprintf("No Heap section\n");
+		free(cnk);
 		return chunks;
 	}
 	const int tcache = rz_config_get_i(core->config, "dbg.glibc.tcache");
@@ -1632,8 +1632,10 @@ RZ_API RzList *GH(rz_heap_arenas_list)(RzCore *core, GHT m_arena, MallocState *m
 		return arena_list;
 	}
 	// main arena
-	GH(rz_heap_update_main_arena)
-	(core, m_arena, ta);
+	if (!GH(rz_heap_update_main_arena)(core, m_arena, ta)) {
+		free(ta);
+		return arena_list;
+	}
 	rz_list_append(arena_list, ta);
 	if (main_arena->GH(next) != m_arena) {
 		ta->GH(next) = main_arena->GH(next);
@@ -1717,13 +1719,6 @@ RZ_API RzList *GH(rz_heap_chunks_list)(RzCore *core, MallocState *main_arena,
 		return chunks;
 	}
 
-	RzConfigHold *hc = rz_config_hold_new(core->config);
-	if (!hc) {
-		free(cnk);
-		free(cnk_next);
-		return chunks;
-	}
-
 	(void)rz_io_read_at(core->io, next_chunk, (ut8 *)cnk, sizeof(GH(RzHeapChunk)));
 	size_tmp = (cnk->size >> 3) << 3;
 	ut64 prev_chunk_addr;
@@ -1785,8 +1780,6 @@ RZ_API RzList *GH(rz_heap_chunks_list)(RzCore *core, MallocState *main_arena,
 		if (tcache) {
 			GH(RTcache) *tcache_heap = GH(tcache_new)(core);
 			if (!tcache_heap) {
-				rz_config_hold_restore(hc);
-				rz_config_hold_free(hc);
 				free(cnk);
 				free(cnk_next);
 				return chunks;
@@ -1973,8 +1966,11 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 	top_data = rz_str_new("");
 	if (mode == RZ_OUTPUT_MODE_JSON) {
 		if (!pj) {
+			rz_cons_newline();
+			free(g);
 			free(top_data);
 			free(main_arena);
+			rz_cons_canvas_free(can);
 			rz_config_hold_restore(hc);
 			rz_config_hold_free(hc);
 			return RZ_CMD_STATUS_ERROR;
@@ -2025,6 +2021,9 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 			rz_cons_printf("f %s %d 0x%" PFMT64x "\n", name, (int)pos->size, (ut64)pos->addr);
 			free(name);
 		} else if (mode == RZ_OUTPUT_MODE_LONG_JSON) { // graph
+			if (node_title) {
+				free(node_title);
+			}
 			node_title = rz_str_newf("  Malloc chunk @ 0x%" PFMT64x " ", (ut64)pos->addr);
 			if (node_data) {
 				free(node_data);
@@ -2067,9 +2066,6 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 			free(node_title);
 		}
 		rz_agraph_print(g);
-		rz_cons_canvas_free(can);
-		rz_config_hold_restore(hc);
-		rz_config_hold_free(hc);
 	}
 	rz_cons_newline();
 	free(g);
@@ -2077,6 +2073,9 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc, 
 	free(top_title);
 	rz_list_free(chunks);
 	free(main_arena);
+        rz_cons_canvas_free(can);
+        rz_config_hold_restore(hc);
+        rz_config_hold_free(hc);
 	return RZ_CMD_STATUS_OK;
 }
 

--- a/librz/core/linux_heap_glibc.c
+++ b/librz/core/linux_heap_glibc.c
@@ -1190,15 +1190,13 @@ static void GH(print_tcache_instance)(RzCore *core, GHT m_arena, MallocState *ma
 	GH(RTcache) * rz_tcache;
 	int count = 0;
 	rz_list_foreach (tcache_list, iter, rz_tcache) {
-		MallocState *prev_arena;
+		RzArenaListItem *arena = rz_list_get_n(arenas_list, count);
 		if (count == 0) {
-			prev_arena = arenas_list->tail->data;
 			rz_cons_printf("Tcache in Main Arena @ ");
 		} else {
-			prev_arena = rz_list_get_n(arenas_list, count - 1);
 			rz_cons_printf("Tcache in Thread Arena @ ");
 		}
-		PRINTF_YA(" 0x%" PFMT64x, (ut64)prev_arena->GH(next));
+		PRINTF_YA(" 0x%" PFMT64x, (ut64)arena->addr);
 		rz_cons_newline();
 		GH(tcache_print)
 		(core, rz_tcache, pj);
@@ -1612,6 +1610,11 @@ static void GH(print_main_arena_bins)(RzCore *core, GHT m_arena, MallocState *ma
 	}
 }
 
+void GH(rz_arena_list_free)(RzArenaListItem *item) {
+	free(item->arena);
+	free(item->type);
+	free(item);
+}
 /**
  * \brief Get a list of MallocState structs for all the arenas
  * \param core RzCore pointer
@@ -1620,7 +1623,7 @@ static void GH(print_main_arena_bins)(RzCore *core, GHT m_arena, MallocState *ma
  * \return RzList pointer for list of MallocState structs of all the arenas
  */
 RZ_API RzList *GH(rz_heap_arenas_list)(RzCore *core, GHT m_arena, MallocState *main_arena) {
-	RzList *arena_list = rz_list_newf(free);
+	RzList *arena_list = rz_list_newf((RzListFree)GH(rz_arena_list_free));
 	MallocState *ta = RZ_NEW0(MallocState);
 	if (!ta) {
 		return arena_list;
@@ -1630,7 +1633,11 @@ RZ_API RzList *GH(rz_heap_arenas_list)(RzCore *core, GHT m_arena, MallocState *m
 		free(ta);
 		return arena_list;
 	}
-	rz_list_append(arena_list, ta);
+	RzArenaListItem *item = RZ_NEW0(RzArenaListItem);
+	item->addr = m_arena;
+	item->type = rz_str_new("Main");
+	item->arena = ta;
+	rz_list_append(arena_list, item);
 	if (main_arena->GH(next) != m_arena) {
 		ta->GH(next) = main_arena->GH(next);
 		while (GH(is_arena)(core, m_arena, ta->GH(next)) && ta->GH(next) != m_arena) {
@@ -1641,12 +1648,20 @@ RZ_API RzList *GH(rz_heap_arenas_list)(RzCore *core, GHT m_arena, MallocState *m
 				return arena_list;
 			}
 			// thread arenas
-			rz_list_append(arena_list, ta);
+			item = RZ_NEW0(RzArenaListItem);
+			item->addr = ta_addr;
+			item->type = rz_str_new("Thread");
+			item->arena = ta;
+			rz_list_append(arena_list, item);
 		}
 	}
 	return arena_list;
 }
 
+void GH(rz_chunk_list_free)(RzHeapChunkListItem *item) {
+	free(item->status);
+	free(item);
+}
 /**
  * \brief Get a list of all the heap chunks in an arena. The chunks are in form of a struct RzHeapChunkListItem
  * \param core RzCore pointer
@@ -1658,7 +1673,7 @@ RZ_API RzList *GH(rz_heap_arenas_list)(RzCore *core, GHT m_arena, MallocState *m
  */
 RZ_API RzList *GH(rz_heap_chunks_list)(RzCore *core, MallocState *main_arena,
 	GHT m_arena, GHT m_state) {
-	RzList *chunks = rz_list_newf(free);
+	RzList *chunks = rz_list_newf((RzListFree)GH(rz_chunk_list_free));
 	if (!core || !core->dbg || !core->dbg->maps) {
 		return chunks;
 	}
@@ -1818,21 +1833,20 @@ RZ_API RzList *GH(rz_heap_chunks_list)(RzCore *core, MallocState *main_arena,
 		rz_io_read_at(core->io, next_chunk, (ut8 *)cnk, sizeof(GH(RzHeapChunk)));
 		size_tmp = (cnk->size >> 3) << 3;
 
-		char *status = "allocated";
+		char *status = rz_str_newlen("allocated", 10);
 		if (fastbin) {
 			if (is_free) {
-				status = "free";
+				strcpy(status, "free");
 			}
 		}
 		if (!(cnk->size & 1)) {
-			status = "free";
+			strcpy(status, "free");
 		}
 		if (tcache) {
 			if (is_free) {
-				status = "free";
+				strcpy(status, "free");
 			}
 		}
-
 		RzHeapChunkListItem *block = malloc(sizeof(RzHeapChunkListItem));
 		block->addr = prev_chunk_addr;
 		block->status = status;
@@ -1860,34 +1874,29 @@ RZ_IPI RzCmdStatus GH(rz_cmd_arena_print_handler)(RzCore *core, int argc, const 
 		return RZ_CMD_STATUS_ERROR;
 	}
 	RzList *arenas_list = GH(rz_heap_arenas_list)(core, m_arena, main_arena);
-	RzListIter *iter, *iter2;
-	MallocState *pos;
-	ut64 addr;
-	rz_list_foreach (arenas_list, iter2, pos) {
-		addr = pos->GH(next);
-	}
-	bool flag = 0;
+	RzListIter *iter;
+	RzArenaListItem *pos;
+	bool flag = false;
 	rz_list_foreach (arenas_list, iter, pos) {
+		MallocState *arena = pos->arena;
 		if (!flag) {
 			flag = true;
 			rz_cons_printf("Main arena  (addr=");
-
 		} else {
 			rz_cons_printf("Thread arena(addr=");
 		}
-		PRINTF_YA("0x%" PFMT64x, (ut64)addr);
+		PRINTF_YA("0x%" PFMT64x, (ut64)pos->addr);
 		rz_cons_printf(", lastRemainder=");
-		PRINTF_YA("0x%" PFMT64x, (ut64)pos->GH(last_remainder));
+		PRINTF_YA("0x%" PFMT64x, (ut64)arena->GH(last_remainder));
 		rz_cons_printf(", top=");
-		PRINTF_YA("0x%" PFMT64x, (ut64)pos->GH(top));
+		PRINTF_YA("0x%" PFMT64x, (ut64)arena->GH(top));
 		rz_cons_printf(", next=");
-		PRINTF_YA("0x%" PFMT64x, (ut64)pos->GH(next));
-		if (pos->attached_threads) {
+		PRINTF_YA("0x%" PFMT64x, (ut64)arena->GH(next));
+		if (arena->attached_threads) {
 			rz_cons_printf(")\n");
 		} else {
 			rz_cons_printf(", free)\n");
 		}
-		addr = (ut64)pos->GH(next);
 	}
 	rz_list_free(arenas_list);
 	free(main_arena);
@@ -2301,4 +2310,39 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_arena_bins_print_handler)(RzCore *core, int ar
 	(core, m_state, main_arena, global_max_fast, bin_format, json);
 	free(main_arena);
 	return RZ_CMD_STATUS_OK;
+}
+
+RZ_API RzList *GH(rz_heap_arena_list_wrapper)(RzCore *core) {
+	GHT m_arena;
+	if (!GH(rz_heap_resolve_main_arena)(core, &m_arena)) {
+		return rz_list_newf(free);
+	}
+	MallocState *main_arena = RZ_NEW0(MallocState);
+	if (!main_arena) {
+		return rz_list_newf(free);
+	}
+	if (!GH(rz_heap_update_main_arena)(core, m_arena, main_arena)) {
+		free(main_arena);
+		return rz_list_newf(free);
+	}
+	return GH(rz_heap_arenas_list)(core, m_arena, main_arena);
+}
+
+RZ_API RzList *GH(rz_heap_chunks_list_wrapper)(RzCore *core, ut64 m_state) {
+	GHT m_arena;
+	if (!GH(rz_heap_resolve_main_arena)(core, &m_arena)) {
+		return rz_list_newf(free);
+	}
+	if (!GH(is_arena)(core, m_arena, m_state)) {
+		return rz_list_newf(free);
+	}
+	MallocState *main_arena = RZ_NEW0(MallocState);
+	if (!main_arena) {
+		return rz_list_newf(free);
+	}
+	if (!GH(rz_heap_update_main_arena)(core, m_state, main_arena)) {
+		free(main_arena);
+		return rz_list_newf(free);
+	}
+	return GH(rz_heap_chunks_list)(core, main_arena, m_arena, m_state);
 }

--- a/librz/include/rz_core.h
+++ b/librz/include/rz_core.h
@@ -775,6 +775,7 @@ RZ_API void rz_core_sysenv_end(RzCore *core, const char *cmd);
 RZ_API void rz_core_recover_vars(RzCore *core, RzAnalysisFunction *fcn, bool argonly);
 
 /* linux_heap_glibc.c */
+// TODO MOVE ME TO DIFFERENT FILE
 RZ_API RzHeapChunk_64 *rz_heap_get_chunk_at_addr_64(RzCore *core, ut64 addr);
 RZ_API RzList *rz_heap_bin_content_list_64(RzCore *core, MallocState *main_arena, int bin_num);
 RZ_API RzList *rz_heap_arenas_list_64(RzCore *core, ut64 m_arena, MallocState *main_arena);
@@ -789,6 +790,10 @@ RZ_API RzList *rz_heap_heap_chunks_list_32(RzCore *core, MallocState *main_arena
 RZ_API bool rz_heap_resolve_main_arena_32(RzCore *core, ut32 *m_arena);
 RZ_API bool rz_heap_update_main_arena_32(RzCore *core, ut32 m_arena, MallocState *main_arena);
 RZ_API RzList *rz_heap_tcache_list_32(RzCore *core, ut32 m_arena, MallocState *main_arena, bool main_thread_only);
+
+/* cmd_linux_heap_glibc.c */
+RZ_API RzList *rz_heap_chunks_list(RzCore *core, ut64 m_arena);
+RZ_API RzList *rz_heap_arenas_list(RzCore *core);
 // XXX dupe from rz_bin.h
 /* bin.c */
 #define RZ_CORE_BIN_ACC_STRINGS          0x001

--- a/librz/include/rz_heap_glibc.h
+++ b/librz/include/rz_heap_glibc.h
@@ -281,6 +281,12 @@ typedef struct rz_heap_chunk_list_item {
 	char *status; /* Status of the chunk, allocated/free/corrupted */
 } RzHeapChunkListItem;
 
+typedef struct rz_arena_list_item {
+	ut64 addr; /* Base addr of the arena */
+	char *type; /* Type of arena, main/thread */
+	MallocState *arena; /* The MallocState for the arena */
+} RzArenaListItem;
+
 #ifdef __cplusplus
 }
 #endif

--- a/librz/include/rz_heap_glibc.h
+++ b/librz/include/rz_heap_glibc.h
@@ -31,7 +31,6 @@ RZ_LIB_VERSION_HEADER(rz_heap_glibc);
 #define FASTBIN_IDX_TO_SIZE(i) ((SZ * 4) + (SZ * 2) * (i - 1))
 #define BITSPERMAP             (1U << BINMAPSHIFT)
 #define BINMAPSIZE             (NBINS / BITSPERMAP)
-#define MAX(a, b)              (((a) > (b)) ? (a) : (b))
 #define NPAD                   -6
 #define TCACHE_MAX_BINS        64
 #define TCACHE_FILL_COUNT      7

